sql/core/src/main/scala/org/apache/spark/sql/Encoder.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import java.lang.reflect.{ParameterizedType, Constructor}
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.api.java.function.{Function => JFunction, Function2 => JFunction2, _}
+import org.apache.spark.sql.catalyst.expressions.{SpecificMutableRow, UnsafeRow}
+import org.apache.spark.sql.catalyst._
+import org.apache.spark.sql.catalyst.expressions.UnsafeRowWriters.UTF8StringWriter
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+import scala.language.implicitConversions
+import scala.reflect._
+import scala.reflect.runtime.universe._
+
+object Encoder {
+  def fromType(genericTypes: Seq[java.lang.reflect.Type]): Encoder[_] = ???
+}
+
+/**
+ * Captures how to encode JVM objects as Spark SQL rows.
+ * TODO: Make unsafe row?
```

Comment:
seems like you might want macros to generate these for case classes and primitives.